// flow-native-token-registry/dist/main/lib/tokenlist.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenListContainer = exports.TokenListProvider = exports.StaticTokenListResolutionStrategy = exports.Strategy = exports.CDNTokenListResolutionStrategy = exports.GitHubTokenListResolutionStrategy = exports.CLUSTER_SLUGS = exports.ENV = void 0;
const cross_fetch_1 = require("cross-fetch");
const flow_mainnet_tokenlist_json_1 = __importDefault(require("./../tokens/flow-mainnet.tokenlist.json"));
const flow_testnet_tokenlist_json_1 = __importDefault(require("./../tokens/flow-testnet.tokenlist.json"));
var ENV;
(function (ENV) {
    ENV["Mainnet"] = "mainnet";
    ENV["Testnet"] = "testnet";
})(ENV = exports.ENV || (exports.ENV = {}));
exports.CLUSTER_SLUGS = {
    mainnet: ENV.Mainnet,
    testnet: ENV.Testnet,
};
class GitHubTokenListResolutionStrategy {
    constructor() {
        this.repositories = {
            [ENV.Mainnet]: [
                'https://raw.githubusercontent.com/FlowFans/flow-token-list/main/src/tokens/flow-mainnet.tokenlist.json',
            ],
            [ENV.Testnet]: [
                'https://raw.githubusercontent.com/FlowFans/flow-token-list/main/src/tokens/flow-testnet.tokenlist.json',
            ],
        };
        this.resolve = (network = ENV.Mainnet) => {
            return queryJsonFiles(this.repositories[network]);
        };
    }
}
exports.GitHubTokenListResolutionStrategy = GitHubTokenListResolutionStrategy;
class CDNTokenListResolutionStrategy {
    constructor() {
        this.repositories = {
            [ENV.Mainnet]: [
                'https://cdn.jsdelivr.net/gh/FlowFans/flow-token-list@main/src/tokens/flow-mainnet.tokenlist.json',
            ],
            [ENV.Testnet]: [
                'https://cdn.jsdelivr.net/gh/FlowFans/flow-token-list@main/src/tokens/flow-testnet.tokenlist.json',
            ],
        };
        this.resolve = (network = ENV.Mainnet) => {
            return queryJsonFiles(this.repositories[network]);
        };
    }
}
exports.CDNTokenListResolutionStrategy = CDNTokenListResolutionStrategy;
const queryJsonFiles = async (files, network = ENV.Mainnet) => {
    const responses = (await Promise.all(files.map(async (repo) => {
        try {
            const response = await (0, cross_fetch_1.fetch)(repo);
            const json = (await response.json());
            return json;
        }
        catch (_a) {
            console.info(`@flowfans/flow-native-token-registry: falling back to static repository.`);
            return network === ENV.Mainnet ? flow_mainnet_tokenlist_json_1.default : flow_testnet_tokenlist_json_1.default;
        }
    })));
    return responses
        .map((tokenlist) => tokenlist.tokens)
        .reduce((acc, arr) => acc.concat(arr), []);
};
var Strategy;
(function (Strategy) {
    Strategy["GitHub"] = "GitHub";
    Strategy["Static"] = "Static";
    Strategy["CDN"] = "CDN";
})(Strategy = exports.Strategy || (exports.Strategy = {}));
class StaticTokenListResolutionStrategy {
    constructor() {
        this.resolve = (network = ENV.Mainnet) => {
            if (network === ENV.Mainnet)
                return flow_mainnet_tokenlist_json_1.default.tokens;
            return flow_testnet_tokenlist_json_1.default.tokens;
        };
    }
}
exports.StaticTokenListResolutionStrategy = StaticTokenListResolutionStrategy;
class TokenListProvider {
    constructor() {
        this.resolve = async (strategy = Strategy.GitHub, network = ENV.Mainnet) => {
            return new TokenListContainer(await TokenListProvider.strategies[strategy].resolve(network));
        };
    }
}
exports.TokenListProvider = TokenListProvider;
TokenListProvider.strategies = {
    [Strategy.GitHub]: new GitHubTokenListResolutionStrategy(),
    [Strategy.Static]: new StaticTokenListResolutionStrategy(),
    [Strategy.CDN]: new CDNTokenListResolutionStrategy(),
};
class TokenListContainer {
    constructor(tokenList) {
        this.tokenList = tokenList;
        this.filterByTag = (tag) => {
            return new TokenListContainer(this.tokenList.filter((item) => (item.tags || []).includes(tag)));
        };
        this.excludeByTag = (tag) => {
            return new TokenListContainer(this.tokenList.filter((item) => !(item.tags || []).includes(tag)));
        };
        this.getList = () => {
            return this.tokenList;
        };
    }
}
exports.TokenListContainer = TokenListContainer;
