// flow-native-token-registry/dist/module/lib/tokenlist.js
import { fetch } from 'cross-fetch';
import tokenlistMainnet from './../tokens/flow-mainnet.tokenlist.json';
import tokenlistTestnet from './../tokens/flow-testnet.tokenlist.json';
export var ENV;
(function (ENV) {
    ENV["Mainnet"] = "mainnet";
    ENV["Testnet"] = "testnet";
})(ENV || (ENV = {}));
export const CLUSTER_SLUGS = {
    mainnet: ENV.Mainnet,
    testnet: ENV.Testnet,
};
export class GitHubTokenListResolutionStrategy {
    repositories = {
        [ENV.Mainnet]: [
            'https://raw.githubusercontent.com/FlowFans/flow-token-list/main/src/tokens/flow-mainnet.tokenlist.json',
        ],
        [ENV.Testnet]: [
            'https://raw.githubusercontent.com/FlowFans/flow-token-list/main/src/tokens/flow-testnet.tokenlist.json',
        ],
    };
    resolve = (network = ENV.Mainnet) => {
        return queryJsonFiles(this.repositories[network]);
    };
}
export class CDNTokenListResolutionStrategy {
    repositories = {
        [ENV.Mainnet]: [
            'https://cdn.jsdelivr.net/gh/FlowFans/flow-token-list@main/src/tokens/flow-mainnet.tokenlist.json',
        ],
        [ENV.Testnet]: [
            'https://cdn.jsdelivr.net/gh/FlowFans/flow-token-list@main/src/tokens/flow-testnet.tokenlist.json',
        ],
    };
    resolve = (network = ENV.Mainnet) => {
        return queryJsonFiles(this.repositories[network]);
    };
}
const queryJsonFiles = async (files, network = ENV.Mainnet) => {
    const responses = (await Promise.all(files.map(async (repo) => {
        try {
            const response = await fetch(repo);
            const json = (await response.json());
            return json;
        }
        catch {
            console.info(`@flowfans/flow-native-token-registry: falling back to static repository.`);
            return network === ENV.Mainnet ? tokenlistMainnet : tokenlistTestnet;
        }
    })));
    return responses
        .map((tokenlist) => tokenlist.tokens)
        .reduce((acc, arr) => acc.concat(arr), []);
};
export var Strategy;
(function (Strategy) {
    Strategy["GitHub"] = "GitHub";
    Strategy["Static"] = "Static";
    Strategy["CDN"] = "CDN";
})(Strategy || (Strategy = {}));
export class StaticTokenListResolutionStrategy {
    resolve = (network = ENV.Mainnet) => {
        if (network === ENV.Mainnet)
            return tokenlistMainnet.tokens;
        return tokenlistTestnet.tokens;
    };
}
export class TokenListProvider {
    static strategies = {
        [Strategy.GitHub]: new GitHubTokenListResolutionStrategy(),
        [Strategy.Static]: new StaticTokenListResolutionStrategy(),
        [Strategy.CDN]: new CDNTokenListResolutionStrategy(),
    };
    resolve = async (strategy = Strategy.GitHub, network = ENV.Mainnet) => {
        return new TokenListContainer(await TokenListProvider.strategies[strategy].resolve(network));
    };
}
export class TokenListContainer {
    tokenList;
    constructor(tokenList) {
        this.tokenList = tokenList;
    }
    filterByTag = (tag) => {
        return new TokenListContainer(this.tokenList.filter((item) => (item.tags || []).includes(tag)));
    };
    excludeByTag = (tag) => {
        return new TokenListContainer(this.tokenList.filter((item) => !(item.tags || []).includes(tag)));
    };
    getList = () => {
        return this.tokenList;
    };
}
